### DISCUSSION TITLE
defaultキーマップ他でボールとマウスボタンが反応しない

### DISCUSSION THREAD

**Discussion by ergoneko** [Yowkees/keyball#71] [Q&A] [answered]
お世話になります．defaultを元にカスタマイズしたキーマップをmakeしたところ，ボールとマウスボタン（kc_BTN1等）が反応しません．解決策を相談します．qmk環境や確認したことは以下です．
【qmk環境】Windows10，MSYS2 MinGW64bit , QMK Firmware 0.14.25  ※2021/10/20に環境を再構築
・mekeコマンド:　qmk compile -kb keyball -km default　　※　make keyball:default 　と同じ操作
・作成されるhexファイルは１つ．これをQMK ToolBox　で，左右のProMicroへそれぞれ書き込んだ

【ボールやマウスボタンの動作状態で確認したこと】
本体やキーマップは共に右ボールです．
１．組立手順に掲載のVIA機能付hexファイル（ball，noball）を左右それぞれに書き込むと，正常動作する．
　→　キーボード本体に問題なさそう．
２．GitHubからDLしたdefaultキーマップを，上記のmake手順で作成したhexファイルを書き込むと，無反応．
　※但し，KC_BTN1 を登録したレイヤのその他のキーや，その他のレイヤは正常動作．

【その他　抱えている問題】
３．defaultから作成したhexファイルでは，右手キーボードとパソコンと接続すると，キーマップが左右逆転する．
　※左手キーボードと接続すると正常配置される．VIA機能付hexファイルでは，右手側を接続しても正常動作．
４．keyballのキーマップの正しいmakeの仕方を教えてください．defaultキーマップで試すとエラーが発生．
・make keyball/rev1/ball:default 　　→　qmk compile -kb keyball/rev1/ball -km default 
・make keyball/rev1/noball:default 　→　qmk compile -kb keyball/rev1/nobal -km default 
一度にたくさん相談してすみません．大変恐縮でございますが，よろしくお願いします．

**Comment by Yowkees** [comment, marked as answer]
お世話になります。
私の解る範囲でご回答させていただきます。

１．make済みのhexファイルで問題なし→ハードはOK
２．make方法が間違っています。rev1以下に/ball /noballを付けてください。
３．私も再度、defaultキーマップにて/ball /noballのhexファイルをmakeして書き込んでみましたが、左右反転の現象は確認できませんでした。USBを右手・左手どちらに接続しても反転はありません。
４．make方法は４に記載いただいた方法で合っています。こちらでdefaultキーマップでmakeしてもエラーは発生しません。
defaultキーマップをベースにカスタマイズいただいているようですので、一度Githubからダウンロードしたそのままのファイルでエラーが出るかご確認をお願いいたします。

ご返答が遅くなり申し訳ありませんでした。ご検討よろしくお願いいたします。

**Reply by ergoneko** [reply]
ご教示有難うございます．qmkのkeyboards 以下のファイルをGitHubから取り直し，再度試したところ，ball ，　noball のhex ファイルを作ることができました．有難うございました．

hexファイルを作れるようになったところで，もう一つ教えてください．画面スクロールモードの速度を少しさげたいのですが，どのようにしたらできるか教えてください．パソコンにatomをインストールしていますので，指定のファイルを開いて編集する環境はあります．よろしくお願いします．

**Reply by Yowkees** [reply]
各キーマップの config.h に下記の１行を追加してください。

#　define TRACKBALL_SCROLL_DIVIDER 30

スクロールモードではトラックボールの移動量を `KEYBALL_SCROLL_DIVIDER` で割るこ
とで程よいスクロール量に調整しています。
デフォルトは `10` で `0` 以下の値を設定するとコンパイルエラーになります。
上記の例では、TRACKBALL_SCROLL_DIVIDERを３倍の30に設定しました。これでスクロール移動量が１／３になるはずです。
ご検討ください。

**Reply by ergoneko** [reply]
ご教示有難うございます．以下の２つの方法で試しましが，スクロール速度が下がったようには感じられず，やり方の確認と，アドバイス文頭の「各キーマップ」が指すものを教えてください．なお，hexファイルはballの方だけ作りました．
実店舗開店直後の忙しい最中と存じ，すみませんがよろしくお願いします．
やり方１．keymaps>default>config.h　ファイルの最下行に，「# define ~」を追加した．数値は30～90を20刻みで試した．
やり方２．keymaps 以下にあるフォルダ（default，test，～，via_Left）の中にある全てのconfig.hファイルの最下行へ「# define ~」を追加した．

**Reply by koron** [reply]
横やりです。

「各キーマップ」 とは ergonekoさんが用意された自前のキーマップのことで、質問の最初の書き込みにあるコレのことです。

> defaultを元にカスタマイズしたキーマップ

仮にこれをergonekoとするならば keymaps/ergoneko/config.h の最下行に以下の内容を追記してください。

```c
#define TRACKBALL_SCROLL_DIVIDER 30
```

その上で以下のようにballおよびnoball版をビルドしてください。

```
make keyball/rev1/ball:erogoneko
make keyball/rev1/noball:erogoneko
```

これで keyball_rev1_ball_ergoneko.hex と keyball_rev1_noball_ergoneko.hex 2つのファームウェアファイルが得られると期待できます。

もしもビルドできずエラーが出るようであればそのエラーの文言を共有していただければ、それをみて対策を検討できます。
またビルドは成功するけど意図通りに動かないようであれば、一連のファイルを共有(レポジトリを公開していただくのがオススメ)していただければやはり対策を検討できます。

**Comment by j1n6** [comment]
Hello, I seem to have similar issues with right/left inversion after building my own firmware layout.
My trackball is on the right side:
 - When plugin to the left side, the keymaping is correct, however the trackball does not work.
 - When plugin to the right side, the key mapping left & right is reversed, trackball is working.

**Reply by j1n6** [reply]
ah, I found the issue. It's the way I built the firmware I you mentioned. 
Using the following would work correctly:
make keyball/rev1/ball:[keymap name]
